src/Binary.hs

```diff
@@ -1,23 +1,34 @@
-module Binary where 
+module Binary (isValid, nums, toBinary, fromBinary) where
+
+import Data.Maybe (fromJust)
 
 -- Checks that the given list is a valid reversed binary representation 
 isValid :: [Int] -> Maybe [Int]
-isValid xs = undefined 
+isValid [] = Nothing
+isValid xs = if last xs == 0 || any (\x -> x /= 0 && x /= 1) xs then Nothing else Just xs
 
 -- Generates an infinite list of reversed binary representations of natural numbers in order
 nums :: [[Int]]
-nums = undefined 
+nums = map (fromJust . toBinary) [0..]
 
 -- Computes the reversed binary representation of a positive number: no leading zeros
 -- toBinary 6 == Just [0,1,1]
 -- toBinary 0 == Just [0]
 -- toBinary (-10) == Nothing 
 toBinary :: Int -> Maybe [Int]
-toBinary n = undefined 
+toBinary n
+  | n < 0 = Nothing
+  | n < 2 = Just [n]
+  | otherwise = Just $ (n `mod` 2) : fromJust (toBinary (n `div` 2))
+
```

Comment:
A more idiomatic way would be to separate validation of the input from the computation of the actual number list.